source/base/mpi.cc

```diff
@@ -753,17 +755,25 @@ namespace Utilities
       int MPI_has_been_started = 0;
       ierr                     = MPI_Initialized(&MPI_has_been_started);
       AssertThrowMPI(ierr);
-      AssertThrow(MPI_has_been_started == 0,
-                  ExcMessage("MPI error. You can only start MPI once!"));
-
-      int provided;
-      // this works like ierr = MPI_Init (&argc, &argv); but tells MPI that
-      // we might use several threads but never call two MPI functions at the
-      // same time. For an explanation see on why we do this see
-      // http://www.open-mpi.org/community/lists/users/2010/03/12244.php
-      int wanted = MPI_THREAD_SERIALIZED;
-      ierr       = MPI_Init_thread(&argc, &argv, wanted, &provided);
-      AssertThrowMPI(ierr);
+
+      if (MPI_has_been_started == 1 && allow_external_initialization)
+        {
+          finalize_mpi = false;
+        }
+      else
+        {
+          AssertThrow(MPI_has_been_started == 0,
+                      ExcMessage("MPI error. You can only start MPI once!"));
```

Comment:
I don't understand why you need to require an extra flag `allow_external_initialization`. For example, the PETSc/SLEPC part can be initalized outside of this function and deal.ii does not complain. What is different for MPI?

Reply:
Currently in `master`, deal.II does *not* allow MPI to be initialized outside of deal.II:
https://github.com/dealii/dealii/blob/3c37cfefb28c7685dafb4434c5b5edf53d845d06/source/base/mpi.cc#L753-L757
I thus assumed that the design goals for this function were that any abnormal usage by the user should result in a fast and hard failure, even though in this case it would have been possible to, e.g., first check the current thread level and see if it is sufficient for deal.II. This makes sense to me - I have used the same approach many times before, assuming that if anything is happening out of the ordinary, the user should probably be informed immediately instead of potentially running an expensive simulation and failing after a couple of hours.

My approach in this PR was therefore motivated by wanting to keep this strict safety against user errors as the default. I therefore made them set a flag explicitly, which indicates to deal.II "I know what I'm doing" but also provides the feedback to the users "if you had to set a flag for this, you're probably in dragon land now".

However, I am not a regular committer to deal.II, so everything I wrote was based only on assumptions. I'd be more than happy to make this work without a flag, if this is so desired. However, from the convo in this PR, I already got the feeling that a very different approach is to be followed for allowing external initialization, so I don't know if this question will not be rendered moot anyway.